pkg/catalog/ingress_test.go

```diff
@@ -1,194 +1,201 @@
 	}
 
-	// Create Bookstore apex Service
-	svc = tests.NewServiceFixture(tests.BookstoreApexService.Name, tests.BookstoreApexService.Namespace, nil)
-	if _, err := kubeClient.CoreV1().Services(tests.BookstoreApexService.Namespace).Create(context.TODO(), svc, metav1.CreateOptions{}); err != nil {
-		GinkgoT().Fatalf("Error creating new Bookstore Apex service", err.Error())
+	type testCase struct {
+		name                    string
+		svc                     service.MeshService
+		ingresses               []*networkingV1beta1.Ingress
+		expectedTrafficPolicies []*trafficpolicy.InboundTrafficPolicy
+		excpectError            bool
 	}
 
-	mockIngressMonitor.EXPECT().GetIngressResources(gomock.Any()).Return(getFakeIngresses(), nil).AnyTimes()
-
-	// Monitored namespaces is made a set to make sure we don't repeat namespaces on mock
-	listExpectedNs := tests.GetUnique([]string{
-		tests.BookstoreV1Service.Namespace,
-		tests.BookbuyerService.Namespace,
-		tests.BookstoreApexService.Namespace,
-	})
-
-	// #1683 tracks potential improvements to the following dynamic mocks
-	mockKubeController.EXPECT().ListServices().DoAndReturn(func() []*corev1.Service {
-		// play pretend this call queries a controller cache
-		var services []*corev1.Service
-
-		for _, ns := range listExpectedNs {
-			// simulate lookup on controller cache
-			svcList, _ := kubeClient.CoreV1().Services(ns).List(context.TODO(), metav1.ListOptions{})
-			for serviceIdx := range svcList.Items {
-				services = append(services, &svcList.Items[serviceIdx])
-			}
-		}
-
-		return services
-	}).AnyTimes()
-	mockKubeController.EXPECT().GetService(gomock.Any()).DoAndReturn(func(msh service.MeshService) *corev1.Service {
-		// simulate lookup on controller cache
-		vv, err := kubeClient.CoreV1().Services(msh.Namespace).Get(context.TODO(), msh.Name, metav1.GetOptions{})
-
-		if err != nil {
-			return nil
-		}
-
-		return vv
-	}).AnyTimes()
-	mockKubeController.EXPECT().IsMonitoredNamespace(tests.BookstoreV1Service.Namespace).Return(true).AnyTimes()
-	mockKubeController.EXPECT().IsMonitoredNamespace(tests.BookstoreV2Service.Namespace).Return(true).AnyTimes()
-	mockKubeController.EXPECT().IsMonitoredNamespace(tests.BookbuyerService.Namespace).Return(true).AnyTimes()
-	mockKubeController.EXPECT().IsMonitoredNamespace(tests.BookwarehouseService.Namespace).Return(true).AnyTimes()
-	mockKubeController.EXPECT().ListMonitoredNamespaces().Return(listExpectedNs, nil).AnyTimes()
-
-	return NewMeshCatalog(mockKubeController, kubeClient, meshSpec, certManager,
-		mockIngressMonitor, stop, cfg, endpointProviders...)
-}
-
-func getFakeIngresses() []*networkingV1beta1.Ingress {
-	return []*networkingV1beta1.Ingress{
+	testCases := []testCase{
 		{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      "ingress-1",
-				Namespace: fakeIngressNamespace,
-				Annotations: map[string]string{
-					constants.OSMKubeResourceMonitorAnnotation: "enabled",
+			name: "Ingress rule with multiple rules and no default backend",
```

Comment:
what does "no default backend" mean here? I see a backend specified in the ingress resource so just curious.

Reply:
Default backend is a backend that you can specify in an ingress resource which can be used by the load balancer to route traffic to when no rules match.

In the below example, there is a default backend (not the same as backends specified for HTTP rules)
```
networkingV1beta1.IngressSpec{
						Backend: &networkingV1beta1.IngressBackend{
							ServiceName: "foo",
							ServicePort: intstr.IntOrString{
								Type:   intstr.Int,
								IntVal: fakeIngressPort,
							},
						},
						Rules: []networkingV1beta1.IngressRule{
```